### PR TITLE
[RF] HS3 RooFit JSON/YML import/export

### PR DESCRIPTION
**Pull request by cburgard**
This is a reworked version of [this older PR](https://github.com/root-project/root/pull/6254).

Support for JSON/YAML I/O of RooFit workspaces is added. For more info, consult the presentation given in the [ROOT PPP meeting](https://indico.cern.ch/e/PPP111).

### CODE REVIEW

**Comment by cburgard** [comment]
Hi everyone!
this is just a quick ping on whether there is any progress in the review of this PR?

**Comment by guitargeek** [comment]
@phsft-bot build

**Comment by guitargeek** [comment]
Hi @cburgard, thanks for the reminder!

There are a few things that need to be done here:

1. We can't merge non-fast forward branches (like branches where you merged `master` in between). Please rebase all your developments on top of the current `master` and update this PR. If can be done in a single commit, because the current commit history of your developments is not so informative anyway :) Once this is done, the CI can also tests the PR (the CI also has problems with non fast-forward branches).
2. We try to have new C++ code formatted nicely with the [ROOT clang-format style](https://github.com/root-project/root/blob/master/.clang-format) can you please iron over new `roofit/hs3` package with `clang-format`, using the linked style file?
3.  Please add a tutorial for the JSON/YML import/export (or more than one if you want) to the [tutorials/roofit](https://github.com/root-project/root/tree/master/tutorials/roofit) directory. It can be in either C++ or Python, as you prefer. We can take care of the translation later.
4. It would be good to have at least two unit tests: one export/import closure test for a binned model, and one for an unbinned model. These tests should be in `roofit/hs3/tests` (compare [`roofit/roofitcore/test`](https://github.com/root-project/root/tree/master/roofit/roofitcore/test) directory structure). If the tests are written in Python, you can add them to the `test/CmakeLists.txt` with the `ROOT_ADD_PYUNITTEST` macro (see for example the [pythonization unit tests](https://github.com/root-project/root/blob/master/bindings/pyroot/pythonizations/test/CMakeLists.txt)). If you need any help on how to write and run the tests don't hesitate to ask!

File: roofit/hs3/src/RYMLParser.cxx
Comment (guitargeek) [inline roofit/hs3/src/RYMLParser.cxx:235]:
```suggestion
  void error_cb(const char* msg, size_t msg_len, void* /*user_data*/){
  void error_cb(const char* msg, size_t msg_len, c4::yml::Location /*not used*/, void* /*user_data*/){
```

This doesn't match the signature expected by the newest rapidyaml (I tried to compile with the `master` branch from https://github.com/biojppm/rapidyaml). Can you just update the signature? Or do you need backwards compatibility with older rapidyaml? I don't think we need that though, since the RooFit HS3 is a new package anyway.

File: roofit/hs3/CMakeLists.txt
Comment (guitargeek) [inline roofit/hs3/CMakeLists.txt:12]:
```suggestion
ROOT_STANDARD_LIBRARY_PACKAGE(RooFitHS3
```
Please prefix the package name with `RooFit` so it's clear what this is (analogous to `RooFitCore` package).

**Comment by guitargeek** [comment]
Hi @cburgard, can you please fix the commit history in this PR by doing a fresh rebase of your commit onto master? LIke this, we can't merge the PR and I think it also can't be testes by the CI (I will try thought).

**Comment by guitargeek** [comment]
@phsft-bot build

**Comment by guitargeek** [comment]
@phsft-bot build

File: roofit/hs3/test/CMakeLists.txt
Comment (guitargeek) [inline roofit/hs3/test/CMakeLists.txt:5]:
```suggestion
ROOT_ADD_PYUNITTEST(roofit-hs3-histfactory-json test_hs3_histfactory_json.py
  COPY_TO_BUILDDIR ${CMAKE_CURRENT_SOURCE_DIR}/test_hs3_histfactory_json_input.root
  )

ROOT_ADD_GTEST(testRooFitHS3 testRooFitHS3.cxx LIBRARIES RooFitCore RooFit RooFitHS3)
```

Please change the names of the test files and unit tests to these conventional names (some file renaming that you need to do will follow from this change).

File: roofit/hs3/CMakeLists.txt
Comment (guitargeek) [inline roofit/hs3/CMakeLists.txt:18]:
```suggestion
    RooFitHS3/JSONInterface.h
    RooFitHS3/RooJSONFactoryWSTool.h
    RooFitHS3/HistFactoryJSONTool.h
    RooFitHS3/JSONParser.h
    RooFitHS3/RYMLParser.h
```

In newer RooFit code (e.g. RooStats, HistFactory, and new multiprocessing libraries), we put the readers in a subdirectory of `inc`. Please move the header files and adapt the CMakeLists.txt (also fix the indentation please).

File: roofit/hs3/CMakeLists.txt
Comment (guitargeek) [inline roofit/hs3/CMakeLists.txt:25]:
```suggestion
    src/RooJSONFactoryWSTool.cxx
    src/HistFactoryJSONTool.cxx
    src/JSONParser.cxx
    src/RYMLParser.cxx
    src/JSONFactories_RooFitCore.cxx
    src/JSONFactories_HistFactory.cxx
```
Fix indentation (to make it consistent with the parameters below).

File: roofit/hs3/inc/HistFactoryJSONTool.h
Comment (guitargeek) [inline roofit/hs3/inc/HistFactoryJSONTool.h:13]:
```suggestion
```
Unused static variable can be removed

File: roofit/hs3/src/RooJSONFactoryWSTool.cxx
Comment (guitargeek) [inline roofit/hs3/src/RooJSONFactoryWSTool.cxx:866]:
```suggestion
       for (RooAbsData * absData : static_range_cast<RooAbsData*>(*dataList)) {
          this->exportData(absData, output);
       }
```
While fixing the shadowing warning (see CI), you can also use the new `static_range_cast`.

File: roofit/hs3/test/argus.cxx
Comment (guitargeek) [inline roofit/hs3/test/argus.cxx:22]:
```suggestion
{
   auto& msg = RooMsgService::instance();
   msg.setGlobalKillBelow(RooFit::WARNING);

   // --- Observable ---
   RooRealVar mes("mes", "m_{ES} (GeV)", 5.20, 5.30);
```
The test is unnecessarily verbose, you can silence everything below `WARNING`.

File: roofit/hs3/test/histfactory.py
Comment (guitargeek) [inline roofit/hs3/test/histfactory.py:5]:
```suggestion
class TestHS3HistFactoryJSON(unittest.TestCase):
```
`HistFactory` is not a good name for a test class.

File: roofit/hs3/test/histfactory.py
Comment (guitargeek) [inline roofit/hs3/test/histfactory.py:4]:
```suggestion

msg = ROOT.RooMsgService.instance()
msg.setGlobalKillBelow(ROOT.RooFit.WARNING)

```
Make the test less verbose

File: roofit/hs3/test/histfactory.py
Comment (guitargeek) [inline roofit/hs3/test/histfactory.py:127]:
```suggestion
```
This `Print()` is not necessary in the test

File: roofit/hs3/test/histfactory.py
Comment (guitargeek) [inline roofit/hs3/test/histfactory.py:178]:
```suggestion
if __name__ == "__main__":
    unittest.main(verbosity=2)
```
This is better here, so all `test_` member functions are automatically detected and run

File: roofit/hs3/test/histfactory.py
Comment (guitargeek) [inline roofit/hs3/test/histfactory.py:158]:
```suggestion
        newpdf.fitTo(
            newdata,
            Strategy=1,
            Minos=newmc.GetParametersOfInterest(),
            GlobalObservables=newmc.GetGlobalObservables(),
            PrintLevel=-1,

        )
```
* no need to use `Save()` if you don't use the result
* make test less verbose by using `PrintLevel=-`
* you can use the new keyword argument Pythonizations to make the test less verbose

File: tutorials/roofit/rf515_hfJSON.js
Comment (guitargeek) [inline tutorials/roofit/rf515_hfJSON.js:1]: Please rename this file to `*.json`. I think the `.js` ending can be misleading for the tutorials directory, which is often checked out by the users. `.js` makes it look like JavaScript :)

File: tutorials/roofit/rf515_hfJSON.py
Comment (guitargeek) [inline tutorials/roofit/rf515_hfJSON.py:1]:
```suggestion
## \file
## \ingroup tutorial_roofit
## \notebook
## <Your tutorial title, ending with a dot>
##
## <Little summary of the tutorial, can be multi-line, ending with a dot>
##
## \macro_code
##
## \date November 2021
## \author Carsten Burgard

import ROOT
```

You need to add this header to the tutorial to the ROOT reference guide picks it up.

See for example https://github.com/root-project/root/blob/master/tutorials/roofit/rf612_recoverFromInvalidParameters.py.

File: tutorials/roofit/rf515_hfJSON.py
Comment (guitargeek) [inline tutorials/roofit/rf515_hfJSON.py:17]:
```suggestion
result = model.GetPdf().fitTo(ws.data("observed"),
                              Save=True,
                              GlobalObservables=model.GetGlobalObservables())
```
Please use the keyword pythonizations in the tutorials so they are promoted (also in the `fitTo` call below)

File: tutorials/roofit/rf515_hfJSON.py
Comment (guitargeek) [inline tutorials/roofit/rf515_hfJSON.py:14]:
```suggestion
model = ws["main_modelConfig"]
```

I think in Python you can just do this (also below).

File: tutorials/roofit/rf515_hfJSON.py
Comment (guitargeek) [inline tutorials/roofit/rf515_hfJSON.py:24]:
```suggestion
# You can again import it if you want and check for closure
```

Please start all these inline comments with a capital letter, like in the other tutorials.

**Review by guitargeek — CHANGES_REQUESTED** [review]
Hi @cburgard, thanks a lot for your recent work on this PR!

I have made a few comments that need to be addressed before we can merge.

You also need to fix the warnings emitted by the CI (just let me know whenever you need to trigger a rebuild on mattermost).

I know this is a bit of work, but at least I refrained from taking a deeper look at the C++ implementation so it could be worse :)

In the best case, we will merge this PR before the end of November to be merged before the ACAT talk.

**Comment by Axel-Naumann** [comment]
Can `rf515_hfJSON.js` be renamed to `rf515_hfJSON.json` as it's not a JavaScript file?

File: tutorials/roofit/rf515_hfJSON.js
Comment (cburgard) [reply tutorials/roofit/rf515_hfJSON.js]: done

File: roofit/hs3/test/testRooFitHS3.cxx
Comment (guitargeek) [inline roofit/hs3/test/testRooFitHS3.cxx:49]: I forgot to ask yesterday: this `wsexportkeys.json` fiel is missing, is that a problem?

**Comment by cburgard** [comment]
> Can `rf515_hfJSON.js` be renamed to `rf515_hfJSON.json` as it's not a JavaScript file?

done

**Comment by cburgard** [comment]
> Hi @cburgard, thanks a lot for your recent work on this PR!
> 
> I have made a few comments that need to be addressed before we can merge.
> 
> You also need to fix the warnings emitted by the CI (just let me know whenever you need to trigger a rebuild on mattermost).
> 
> I know this is a bit of work, but at least I refrained from taking a deeper look at the C++ implementation so it could be worse :)
> 
> In the best case, we will merge this PR before the end of November to be merged before the ACAT talk.

I went through all of your comments and addressed them the best I could. I hope I didn't miss any of them.

File: roofit/hs3/test/testRooFitHS3.cxx
Comment (cburgard) [reply roofit/hs3/test/testRooFitHS3.cxx:49]: Oh, I never added that? Well, I'll need to add that for sure. Question is where it should go. It should be accessible via $ROOTSYS ideally.

Comment (guitargeek) [reply roofit/hs3/test/testRooFitHS3.cxx:49]: Ok I'll think about where is should go, I'll trigger the CI in the meantime

**Comment by guitargeek** [comment]
@phsft-bot build

File: roofit/hs3/src/HistFactoryJSONTool.cxx
Comment (guitargeek) [inline roofit/hs3/src/HistFactoryJSONTool.cxx:250]:
```suggestion
#ifdef R__HAS_RYML    
void RooStats::HistFactory::JSONTool::PrintYAML(std::ostream &os)    
{    
   TRYMLTree p;    
   auto &n = p.rootnode();    
   n.set_map();    
   this->Export(n);    
   n.writeYML(os);    
}    
#else    
void RooStats::HistFactory::JSONTool::PrintYAML(std::ostream & /*os*/)       
{    
   std::cerr << "YAML export only support with rapidyaml!" << std::endl;    
}    
#endif    

```
To suppress the warning about the unused `os`, we have to also wrap the function in the signature into the `ifdef` and comment out `os` in the no RYML case.

File: roofit/hs3/src/JSONFactories_HistFactory.cxx
Comment (guitargeek) [inline roofit/hs3/src/JSONFactories_HistFactory.cxx:477]:
```suggestion
```
Remove unused variable `pip`.

File: roofit/hs3/src/JSONParser.cxx
Comment (guitargeek) [inline roofit/hs3/src/JSONParser.cxx:45]:
```suggestion
   Impl(const std::string &k) : _key(k) {}
   virtual ~Impl() = default;
```
We need a virtual destructor for this virtual class

File: roofit/hs3/src/JSONParser.cxx
Comment (guitargeek) [inline roofit/hs3/src/JSONParser.cxx:63]:
```suggestion
   virtual nlohmann::json &get() overrride { return node; }
```

File: roofit/hs3/test/testRooFitHS3.cxx
Comment (guitargeek) [inline roofit/hs3/test/testRooFitHS3.cxx:47]:
```suggestion
```

Remove this stuff irrelevant to the test (also to silence the warning about unused variable `data`).

File: roofit/hs3/test/testRooFitHS3.cxx
Comment (guitargeek) [inline roofit/hs3/test/testRooFitHS3.cxx:53]:
```suggestion
   tool.exportJSON("argus.json");
```
Let's be consistent in using `.json` for all JSON files.

**Review by guitargeek — CHANGES_REQUESTED** [review]
Hi @cburgard, I have to request a few more changes to make the CI warnings go away.

In addition to implementing these comment, there are still some things that need to be done:

1. We need to find a way to silence these warnings: 
   ```
   [2021-11-26T13:49:41.664Z] /mnt/build/workspace/root-pullrequests-build/root/roofit/hs3/src/JSONFactories_RooFitCore.cxx:41:6: warning: ‘{anonymous}::_rooprodpdffactory’ defined but not used [-Wunused-variable]
   ```
   I don't know an easy way to fix this at the moment (the obvious fix doesn't work :)), maybe you have an idea in the meantime before I have time to come back to this.

2. We need to do something about this `wsexportkeys.json` file. I don't know what at the moment, but I'll try out some things. Can you send me this file privately on mattermost?

Thanks!

**Comment by guitargeek** [comment]
@phsft-bot build

**Comment by guitargeek** [comment]
@phsft-bot build

**Comment by guitargeek** [comment]
@phsft-bot build

**Comment by guitargeek** [comment]
@phsft-bot build

**Comment by guitargeek** [comment]
@phsft-bot build

**Comment by guitargeek** [comment]
@phsft-bot build

**Review by guitargeek — APPROVED** [review]
Thank you very much @cburgard for this contribution! Now I'm happy with the way that it's integrated in the build system, and thanks for providing also a tutorial and unit tests!

There is one thing I would like to see in a follow-up PR, related to this line in the tutorial:
https://github.com/root-project/root/pull/8944/files#diff-283d6538b675e263cc0452b64b49dba291f37505dce63cbcd703815a1f5cf1f6R47

When using the conversion tool, the export keys that are integrated into ROOT should be loaded by default. I will take care of this and also about further interface and code improvements before the release.

**Comment by guitargeek** [comment]
@phsft-bot build

File: roofit/hs3/inc/RooFitHS3/RooJSONFactoryWSTool.h
Comment (lmoneta) [inline roofit/hs3/inc/RooFitHS3/RooJSONFactoryWSTool.h:147]: Hi, 
 Is this function really needed ? 
 One can always use the export to string and then do:
 ```
 std::ostream out;
out << exportJSONtoString();
```
This will probably avoid the failure in the current test

**Review by lmoneta — APPROVED** [review]
Thank you for this nice contribution !
Just a small comment, if possible try to keep the interface simple, avoiding possible duplications.

**Comment by guitargeek** [comment]
@phsft-bot build

File: roofit/hs3/inc/RooFitHS3/RooJSONFactoryWSTool.h
Comment (guitargeek) [reply roofit/hs3/inc/RooFitHS3/RooJSONFactoryWSTool.h:147]: Removing this test function doesn't fix the test failure, it was an internal error in the JSONParser that I now fixed.

I agree that in general the interface needs to be improved, which we can do in a follow-up PR before the release.

**Pull request merged**
